### PR TITLE
Use Java string hash seeding and xorshift random numbers.

### PR DESCRIPTION
**Pull request by karalabe**
This PR fixes up the blockies random number generation. It contains three essential patches:

 - Instead of relying on a purely XOR based seed generation (which cannot provide more entropy than the input characters have), use a simple string hashing algorithm (based on [Java's `String.hashCode()`](http://www.docjar.com/html/api/java/lang/String.java.html#1491), expanded to 4 32bit values) to properly saturate all bits and not only those present in the input seed string.
 - The original blockies seed number was a 16 bit integer, so a maximum of 65536 unique avatars could be generated with it. Per the [birthday paradox](https://en.wikipedia.org/wiki/Birthday_problem), even if saturated it's quite easy to find collisions in this space. The new seed introduced to fix this is 4 x 32 = 128 total bits of input entropy.
 - The random number generator itself was based on some strange idea of trigonometric functions. This was swapped out to a proven (and very fast) real PRNG, namely [George Marsaglia' XorShift](https://en.wikipedia.org/wiki/Xorshift).

These modification in total should provide about 2^128 unique avatars, at least as long as not limited by the avatar generation code itself.

### CODE REVIEW

**Comment by Gustav-Simonsson** [comment]
@karalabe Could we expand the seed size to 256 bits? Ethereum, Bitcoin and many other blockchain systems uses 160 bit addresses - would be nice to get all of that entropy into the seed.

Also, while `XorShift` seems like a good choice, is there no standard PRNG in JS stdlibs or so that can be used here? Always nice to reuse existing libs for this type of algos unless there is significant benefit to introducing a new one.

**Comment by karalabe** [comment]
@Gustav-Simonsson Xorshift uses 4 32-bit integers as the seed and state space, so unless we pick a different PRNG algorithm, that's pretty much what we've got to work with :) Also given that blockies is a more general avatar library than Ethereum, I'd rather not aim for such security for displaying some pixels :D 

Regarding build in random numbers, JavaScript does have a built in generator, but that is global across all invocations AFAIK, so generating multiple avatars concurrently or even having something else run in the background would change the end result. Don't know if there's a way to create a local random generator that's independent of the global one.

File: blockies.js [hunk 1]
```diff
@@ -1,18 +1,25 @@
...
-		var n = (Math.sin(randseed++) + 1) / 2; // Move range from -1 - 1 to 0 - 1
-		var r = n * 10000; // Remove first few digits since they tend to have low entropy
-		n = r - Math.floor(r);
-		return n;
+		var t = randseed[0] ^ (randseed[0] << 11);
+
+		randseed[0] = randseed[1];
+		randseed[1] = randseed[2];
+		randseed[2] = randseed[3];
+		randseed[3] = (randseed[3] ^ (randseed[3] >> 19) ^ t ^ (t >> 8));
```
Comment (alexvandesande) [inline blockies.js]: Black magic goes here.

Comment (karalabe) [inline blockies.js]: http://www.jstatsoft.org/v08/i14/paper ;)

**Comment by Gustav-Simonsson** [comment]
@karalabe I don't think concurrent generation of avatars is a concern :) I also don't think there is a drawback with something else making use of the global PRNG, if it has good entropy. However, it looks like `Math.random` is automatically seeded and cannot later be explicitly seeded, so that would be a good reason for not using it here.

Looks like `XorShift` can easily be extended to larger bit sizes - it's wp article has an example with 1024 bits.

**Comment by karalabe** [comment]
@Gustav-Simonsson I think there are some issues with 64 bit numbers. AFAIK, JavaScript cannot really handle them well, so the only way is to split them into 2 32bit ones, do the black magic and convert them back. I saw some similar approach in another js xorshift implementation lib. If you're up for the challenge, I'm all in for raising the security even more :D

**Pull request merged**